### PR TITLE
search: add select:file.directory

### PR DESCRIPTION
**Pull request by rvantonder**
Adds `select:file.directory` support to `select`. All this does is get `dirname` of the file paths of file path results, set it, and deduplicates. The frontend view works as expected when clicking on path links. This is a customer request, and I think ~campaigns~ batch changes would get use out of t.

This does make `select` potentially side-effecting. I think it is an OK side-effect for `file.directory`. We could separate out select functionality that is side-effect free, but I don't think there's a big benefit of doing that right now.

### CODE REVIEW

File: dev/gqltest/search_test.go [hunk 1]
```diff
@@ -1105,9 +1105,9 @@ func testSearchClient(t *testing.T, client searchClient) {
 				counts: counts{File: 1},
 			},
 			{
-				name:   `or statement merges file`,
-				query:  `repo:go-diff HunkNoChunksize or ParseHunksAndPrintHunks select:file`,
-				counts: counts{File: 1},
```
Comment (rvantonder) [inline dev/gqltest/search_test.go:1110]: Lol whoops I shouldn't have deleted this test, meant to copy.

File: internal/search/result/file.go [hunk 2]
```diff
@@ -74,12 +76,21 @@ func (fm *FileMatch) Select(path filter.SelectPath) Match {
 	case filter.File:
 		fm.LineMatches = nil
 		fm.Symbols = nil
+		if len(selectPath) > 1 && selectPath[1] == "directory" {
+			fm.Path = path.Dir(fm.Path)
+			if fm.Path == "." {
+				// When the path contains no directory, we get `.` which is
```
Comment (camdencheek) [inline internal/search/result/file.go:82]: For dirs, should we just always end with a `/` to make it more clear?

**Review by camdencheek — APPROVED** [review]

File: internal/search/result/file.go (hunk 2 above)
Comment (rvantonder) [reply internal/search/result/file.go:82]: That's a good idea! Let me check if it works :-)

**Comment by camdencheek** [comment]
> This does make select potentially side-effecting.

Yeah, I agree that it doesn't seem worth worrying about right now. worst case, we can just make a shallow copy, but I generally expect select to take ownership of the values passed to it.

File: internal/search/result/file.go (hunk 2 above)
Comment (rvantonder) [reply internal/search/result/file.go:82]: Think it works. Potentially a little visually awkward, but let's see how it goes.

File: internal/search/result/file.go [hunk 3]
```diff
@@ -74,12 +76,15 @@ func (fm *FileMatch) Select(path filter.SelectPath) Match {
 	case filter.File:
 		fm.LineMatches = nil
 		fm.Symbols = nil
+		if len(selectPath) > 1 && selectPath[1] == "directory" {
+			fm.Path = path.Dir(fm.Path) + "/" // Add trailing slash for clarity.
```
Comment (keegancsmith) [inline internal/search/result/file.go:80]: you may want to throw in a `path.Clean` or adjust he logic a bit. Here is the output of your current logic:
```
foo -> ./
foo/ -> foo/
/foo -> //
/foo/ -> /foo/
```

https://play.golang.org/p/GjZB0_y5FXi

**Review by keegancsmith — APPROVED** [review]
nice. there is an issue I pointed out inline.

File: internal/search/result/file.go (hunk 3 above)
Comment (rvantonder) [reply internal/search/result/file.go:80]: What a handy function that `path.Clean`! I think the only problematic output from the examples above is `//`, but I think we don't ever return file paths that start with `/`. I think a call to `path.Clean` covers all the cases!

**Pull request merged**